java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java

```diff
@@ -0,0 +1,476 @@
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multiset;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.sonar.java.cfg.CFG;
+import org.sonar.java.model.JavaTree;
+import org.sonar.java.se.ConstraintManager.NullConstraint;
+import org.sonar.java.se.checks.ConditionAlwaysTrueOrFalseCheck;
+import org.sonar.java.se.checks.NullDereferenceCheck;
+import org.sonar.java.se.checks.SECheck;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.ArrayAccessExpressionTree;
+import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.BinaryExpressionTree;
+import org.sonar.plugins.java.api.tree.BlockTree;
+import org.sonar.plugins.java.api.tree.ConditionalExpressionTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.ForStatementTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.IfStatementTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.NewArrayTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+import org.sonar.plugins.java.api.tree.WhileStatementTree;
+
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class ExplodedGraphWalker extends BaseTreeVisitor {
+
+  /**
+   * Arbitrary number to limit symbolic execution.
+   */
+  private static final int MAX_STEPS = 20000;
+  private static final Logger LOG = LoggerFactory.getLogger(ExplodedGraphWalker.class);
+  private static final Set<String> THIS_SUPER = ImmutableSet.of("this", "super");
+
+  private static final boolean DEBUG_MODE_ACTIVATED = false;
+  private static final int MAX_EXEC_PROGRAM_POINT = 2;
+  private final ConditionAlwaysTrueOrFalseCheck alwaysTrueOrFalseChecker;
+  private ExplodedGraph explodedGraph;
+  private Deque<ExplodedGraph.Node> workList;
+  private ExplodedGraph.Node node;
+  public ExplodedGraph.ProgramPoint programPosition;
+  public ProgramState programState;
+
+  private CheckerDispatcher checkerDispatcher;
+
+  @VisibleForTesting
+  int steps;
+  ConstraintManager constraintManager;
+
+  public static class MaximumStepsReachedException extends RuntimeException {
+    public MaximumStepsReachedException(String s) {
+      super(s);
+    }
+  }
+
+  public ExplodedGraphWalker(JavaFileScannerContext context) throws MaximumStepsReachedException {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove the declaration of thrown exception 'org.sonar.java.se.ExplodedGraphWalker$MaximumStepsReachedException' which is a runtime exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)